packages/desktop/src/services/FileDownloadServiceElectron.ts

```diff
@@ -317,12 +419,92 @@ export default class FileDownloadServiceElectron
         });
     }
 
-    public async prepareHttpResourceForDownload(url: string, postBody: string): Promise<Blob> {
-        const response = await this.rawPost<string>(url, postBody);
-        return new Blob([response], { type: "application/json" });
+    private async downloadS3Directory(
+        fileInfo: FileInfo,
+        downloadRequestId: string,
+        onProgress?: (transferredBytes: number) => void,
+        destination?: string
+    ): Promise<DownloadResult> {
+        const { bucket, key, region } = this.parseS3Url(fileInfo.path);
+        const destinationDir = destination || (await this.getDefaultDownloadDirectory());
+        const fullDestinationDir = path.join(destinationDir, fileInfo.name);
+
+        try {
+            // Ensure the destination directory exists
+            fs.mkdirSync(fullDestinationDir, { recursive: true });
+
+            const keys = await this.listS3Objects(bucket, key, region);
+
+            if (keys.length === 0) {
+                throw new Error("No files found in the specified S3 directory.");
+            }
+
+            for (const fileKey of keys) {
+                if (!fileKey) {
+                    console.warn(`Encountered null or undefined file key. Skipping.`);
+                    continue;
+                }
+
+                const relativePath = path.relative(key, fileKey);
+                const destinationPath = path.join(fullDestinationDir, relativePath);
+
+                // Ensure the subdirectories exist
+                fs.mkdirSync(path.dirname(destinationPath), { recursive: true });
+
+                const fileUrl = `https://${bucket}.s3.${region}.amazonaws.com/${encodeURIComponent(
+                    fileKey
+                )}`;
+
+                await this.downloadS3File(fileUrl, destinationPath, onProgress);
+            }
+
+            return {
+                downloadRequestId: fileInfo.id,
+                msg: `Successfully downloaded ${fileInfo.path} to ${fullDestinationDir}`,
+                resolution: DownloadResolution.SUCCESS,
+            };
+        } catch (err: unknown) {
+            console.error(`Failed to download directory: ${err}`);
+            throw new DownloadFailure(
+                `Failed to download directory: ${(err as Error).message}`,
+                downloadRequestId
+            );
+        }
     }
 
-    public cancelActiveRequest(downloadRequestId: string) {
+    private async downloadS3File(
```

Comment:
This seems generic, could it be `downloadFile`?

Reply:
we make the distinction above of downloading a Http file. This also downloads a Http file just without a lot of the wrapping since we have to do a large amount of then for zarr. would something like downloadSimpleHttpFile be acceptable?

Reply:
If they both download a file over http what is the difference? Could we re-use the other function?